clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,952 @@
+            try {
+                if (isTerminatingState()) {
+                    return Optional.empty();
+                }
+
+                if (!maybeSetState(ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS)) {
+                    return Optional.empty();
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+            AbstractRequest.Builder<?> requestBuilder = new GetTelemetrySubscriptionsRequest.Builder(
+                new GetTelemetrySubscriptionsRequestData().setClientInstanceId(clientInstanceId), true);
+            return Optional.of(requestBuilder);
+        }
+
+        private Optional<Builder<?>> createPushRequest(ClientTelemetrySubscription localSubscription) {
+            if (localSubscription == null) {
+                log.warn("Telemetry state is {} but subscription is null; not sending telemetry", state);
+                if (!maybeSetState(ClientTelemetryState.SUBSCRIPTION_NEEDED)) {
+                    log.error("Unable to transition state after failed create push telemetry from state {}", state);
+                }
+                return Optional.empty();
+            }
+
+            // Don't send a push request if we don't have the collector initialized. Re-attempt
+            // the push on the next interval.
+            if (kafkaMetricsCollector == null) {
+                log.warn("Cannot make telemetry request as collector is not initialized");
+                // Update last accessed time for push request to be retried on next interval.
+                updateErrorResult(localSubscription.pushIntervalMs, time.milliseconds());
+                return Optional.empty();
+            }
+
+            boolean terminating;
+            lock.writeLock().lock();
+            try {
+                // We've already been terminated, or we've already issued our last push, so we
+                // should just exit now.
+                if (state == ClientTelemetryState.TERMINATED || state == ClientTelemetryState.TERMINATING_PUSH_IN_PROGRESS) {
+                    return Optional.empty();
+                }
+
+                // Check the *actual* state (while locked) to make sure we're still in the state
+                // we expect to be in.
+                terminating = state == ClientTelemetryState.TERMINATING_PUSH_NEEDED;
+                if (!maybeSetState(terminating ? ClientTelemetryState.TERMINATING_PUSH_IN_PROGRESS : ClientTelemetryState.PUSH_IN_PROGRESS)) {
+                    return Optional.empty();
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+            byte[] payload;
+            try (MetricsEmitter emitter = new ClientTelemetryEmitter(localSubscription.selector(), localSubscription.deltaTemporality())) {
+                emitter.init();
+                kafkaMetricsCollector.collect(emitter);
+                payload = createPayload(emitter.emittedMetrics());
```

Comment:
If you can't create a payload due to some exception, do you return an empty Optional? If so can we move the catch{ } in createPayload here?

Reply:
Make sense, done.